File: frontend/js/trends.js
/**
 * Trends API client module.
 */
const TrendsService = {
    API_BASE: '/api',
    currentTrendData: null,

    /**
     * Get session ID from localStorage
     */
    getSessionId() {
        return localStorage.getItem('osc_finops_session_id');
    },

    /**
     * Build headers with session ID
     */
    getHeaders() {
        const headers = {
            'Content-Type': 'application/json'
        };
        const sessionId = this.getSessionId();
        if (sessionId) {
            headers['X-Session-ID'] = sessionId;
        }
        return headers;
    },

    /**
     * Submit async trend calculation job.
     * 
     * Note: Projection happens automatically when to_date extends beyond yesterday.
     * If budget_id is provided, projected periods are aligned to budget boundaries.
     * 
     * @param {Object} params - Query parameters
     * @param {string} params.from_date - Start date (YYYY-MM-DD)
     * @param {string} params.to_date - End date (YYYY-MM-DD). If after yesterday, projection will occur.
     * @param {string} [params.granularity] - Granularity: 'day', 'week', 'month'
     * @param {string} [params.region] - Filter by region
     * @param {string} [params.resource_type] - Filter by resource type
     * @param {boolean} [params.force_refresh] - Force cache refresh
     * @param {string} [params.budget_id] - Budget ID for period boundary alignment (optional)
     * @returns {Promise<Object>} Job submission response with job_id
     */
    async submitTrendsJob(params) {
        try {
            const response = await fetch(`${this.API_BASE}/trends/async`, {
                method: 'POST',
                headers: {
                    ...this.getHeaders(),
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(params)
            });
            
            const data = await response.json();
            
            if (!response.ok) {
                throw new Error(data.error?.message || `HTTP ${response.status}: ${response.statusText}`);
            }
            
            if (data.success) {
                return data;
            } else {
                throw new Error(data.error?.message || 'Failed to submit trend job');
            }
        } catch (error) {
            console.error('Submit trends job error:', error);
            throw error;
        }
    },

    /**
     * Get job status.
     * @param {string} job_id - Job identifier
     * @returns {Promise<Object>} Job status and result
     */
    async getJobStatus(job_id) {
        try {
            const response = await fetch(`${this.API_BASE}/trends/jobs/${job_id}`, {
                method: 'GET',
                headers: this.getHeaders()
            });
            
            const data = await response.json();
            
            if (!response.ok) {
                throw new Error(data.error?.message || `HTTP ${response.status}: ${response.statusText}`);
            }
            
            if (data.success) {
                return data;
            } else {
                throw new Error(data.error?.message || 'Failed to get job status');
            }
        } catch (error) {
            console.error('Get job status error:', error);
            throw error;
        }
    },

    /**
     * Poll job status until complete.
     * @param {string} job_id - Job identifier
     * @param {Function} onProgress - Callback for progress updates (progress, estimated_remaining)
     * @param {number} interval - Polling interval in milliseconds (default: 2000)
     * @returns {Promise<Object>} Final job result
     */
    async pollJobStatus(job_id, onProgress = null, interval = 2000) {
        return new Promise((resolve, reject) => {
            const poll = async () => {
                try {
                    const status = await this.getJobStatus(job_id);
                    
                    // Call progress callback if provided
                    if (onProgress && status.progress !== undefined) {
                        onProgress(status.progress, status.estimated_time_remaining);
                    }
                    
                    if (status.status === 'completed') {
                        resolve(status);
                    } else if (status.status === 'failed') {
                        reject(new Error(status.error || 'Job failed'));
                    } else {
                        // Continue polling
                        setTimeout(poll, interval);
                    }
                } catch (error) {
                    reject(error);
                }
            };
            
            poll();
        });
    },

};

File: backend/api/trends.py
"""Trends API endpoints."""
import threading
from typing import Optional
from flask import Blueprint, request, jsonify
from datetime import datetime

from backend.services.trend_service import calculate_trends_async
from backend.utils.date_validators import validate_date_range
from backend.services.job_queue import job_queue
from backend.config.settings import SUPPORTED_REGIONS
from backend.utils.errors import APIError
from backend.middleware.auth_middleware import require_auth


trends_bp = Blueprint('trends', __name__)


from backend.utils.session_helpers import get_account_id_from_session


@trends_bp.route('/trends/async', methods=['POST'])
@require_auth
def submit_trends_job():
    """
    Submit async trend calculation job.
    Authentication required.
    
    Request body:
        - from_date: Start date (required, ISO format: YYYY-MM-DD)
        - to_date: End date (required, ISO format: YYYY-MM-DD)
        - granularity: "day", "week", or "month" (optional, default: "day")
        - region: Filter by region (optional, uses session region if not provided)
        - resource_type: Filter by resource type (optional)
        - force_refresh: Force refresh cache (optional: true/false)
    
    Returns:
        Job ID and initial status
    """
    session = getattr(request, 'session', None)
    if not session:
        raise APIError("Authentication required", status_code=401)
    
    # Get request body
    data = request.get_json() or {}
    from_date = data.get('from_date')
    to_date = data.get('to_date')
    granularity = data.get('granularity', 'day')
    region = data.get('region')
    resource_type = data.get('resource_type')
    force_refresh = data.get('force_refresh', False)
    
    # Validate required parameters
    if not from_date or not to_date:
        raise APIError("from_date and to_date are required", status_code=400)
    
    # Validate date range using centralized validator
    is_valid, error_msg = validate_date_range(from_date, to_date, granularity)
    if not is_valid:
        raise APIError(error_msg, status_code=400)
    
    # Validate granularity
    if granularity not in ['day', 'week', 'month']:
        raise APIError("granularity must be 'day', 'week', or 'month'", status_code=400)
    
    # Validate region if provided
    if region and region not in SUPPORTED_REGIONS:
        raise APIError(
            f"Unsupported region: {region}. Supported regions: {', '.join(SUPPORTED_REGIONS)}",
            status_code=400
        )
    
    # Get account_id from session
    account_id = get_account_id_from_session()
    if not account_id:
        raise APIError("Could not retrieve account information", status_code=401)
    
    # Use session region if region not specified
    query_region = region or session.region
    
    # Create job
    job_id = job_queue.create_job(
        job_type="trends",
        metadata={
            "from_date": from_date,
            "to_date": to_date,
            "granularity": granularity,
            "region": query_region,
            "resource_type": resource_type
        }
    )
    
    # Start background thread to process job
    def process_job():
        """Process job in background thread."""
        try:
            # Set status to processing
            job_queue.set_status(job_id, "processing")
            
            # Progress callback
            def progress_callback(progress: int, estimated_remaining: Optional[int]):
                job_queue.set_progress(job_id, progress, estimated_remaining)
            
            # Calculate trends
            result = calculate_trends_async(
                job_id=job_id,
                access_key=session.access_key,
                secret_key=session.secret_key,
                region=query_region,
                account_id=account_id,
                from_date=from_date,
                to_date=to_date,
                granularity=granularity,
                resource_type=resource_type,
                force_refresh=force_refresh,
                progress_callback=progress_callback
            )
            
            # Set result
            job_queue.set_result(job_id, result)
            
        except Exception as e:
            # Set error
            job_queue.set_error(job_id, str(e))
    
    # Start background thread
    thread = threading.Thread(target=process_job, daemon=True)
    thread.start()
    
    return jsonify({
        "success": True,
        "job_id": job_id,
        "status": "pending",
        "message": "Job submitted successfully"
    }), 202


@trends_bp.route('/trends/jobs/<job_id>', methods=['GET'])
@require_auth
def get_job_status(job_id):
    """
    Get job status and results.
    Authentication required.
    
    Path parameters:
        - job_id: Job identifier
    
    Returns:
        Job status, progress, result (if completed), or error (if failed)
    """
    session = getattr(request, 'session', None)
    if not session:
        raise APIError("Authentication required", status_code=401)
    
    # Get job
    job = job_queue.get_job(job_id)
    
    if not job:
        raise APIError("Job not found", status_code=404)
    
    # Verify job belongs to current user (optional security check)
    # For now, we'll allow any authenticated user to check any job
    
    # Format response
    response_data = {
        "success": True,
        "job_id": job_id,
        "status": job["status"],
        "progress": job["progress"],
        "estimated_time_remaining": job.get("estimated_time_remaining"),
        "created_at": job["created_at"].isoformat() if isinstance(job["created_at"], datetime) else str(job["created_at"]),
        "updated_at": job["updated_at"].isoformat() if isinstance(job["updated_at"], datetime) else str(job["updated_at"])
    }
    
    # Add result if completed
    if job["status"] == "completed" and job["result"]:
        response_data["result"] = job["result"]
    
    # Add error if failed
    if job["status"] == "failed" and job["error"]:
        response_data["error"] = job["error"]
    
    return jsonify(response_data), 200

